lucene/core/src/test/org/apache/lucene/search/TestConstantScoreScorer.java

```diff
@@ -0,0 +1,104 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.search;
+
+import java.io.IOException;
+
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.RandomIndexWriter;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.LuceneTestCase;
+
+import static org.apache.lucene.search.BooleanClause.Occur;
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+
+public class TestConstantScoreScorer extends LuceneTestCase {
+  private static final String FIELD = "f";
+  private static final String[] VALUES = new String[]{"foo", "bar", "foo bar", "azerty"};
+
+  public void testBasics() throws Exception {
+    try (final TestConstantScoreScorerIndex index = new TestConstantScoreScorerIndex()) {
+      BooleanQuery query = new BooleanQuery.Builder()
+          .add(new ConstantScoreQuery(new TermQuery(new Term(FIELD, "foo"))), Occur.MUST)
+          .build();
+
+      final ConstantScoreScorer scorer = index.constantScoreScorer(query, 1f);
+      assertThat(scorer.iterator().nextDoc(), equalTo(0));
+      assertThat(scorer.score(), is(equalTo(1f)));
+      assertThat(scorer.iterator().nextDoc(), equalTo(2));
+      assertThat(scorer.score(), is(equalTo(1f)));
+      assertThat(scorer.iterator().nextDoc(), equalTo(NO_MORE_DOCS));
+    }
+  }
+
+  public void testWithMinCompetitiveScoreSet() throws Exception {
+    try (final TestConstantScoreScorerIndex index = new TestConstantScoreScorerIndex()) {
+      BooleanQuery query = new BooleanQuery.Builder()
+          .add(new ConstantScoreQuery(new TermQuery(new Term(FIELD, "foo"))), Occur.MUST)
+          .build();
+
+      final ConstantScoreScorer scorer = index.constantScoreScorer(query, 1f);
+      assertThat(scorer.iterator().nextDoc(), equalTo(0));
+      assertThat(scorer.score(), is(equalTo(1f)));
+      scorer.setMinCompetitiveScore(2f);
+      assertThat(scorer.iterator().nextDoc(), equalTo(NO_MORE_DOCS));
+    }
+  }
+
+  static class TestConstantScoreScorerIndex implements AutoCloseable {
+
+    private final Directory directory;
+    private final RandomIndexWriter writer;
+    private final IndexReader reader;
+
+    TestConstantScoreScorerIndex() throws IOException {
+      directory = newDirectory();
+
+      writer = new RandomIndexWriter(random(), directory);
+      for (String VALUE : VALUES) {
+        Document doc = new Document();
+        doc.add(newTextField(FIELD, VALUE, Field.Store.YES));
+        writer.addDocument(doc);
+      }
+      reader = writer.getReader();
```

Comment:
The random index writer can commit segments randomly so you need to force merge to 1 segments to ensure that all documents are in a single segment:
```
writer.forceMerge(1);
```

Reply:
Done in 8d7d5aefd477dc86a35cc3eb137a8a45534db285, thank you.

Could you please explain me why docs have to be in a single segment to ensure this test passes?

Reply:
Because you check only one segment in the test:
https://github.com/apache/lucene-solr/blob/8d7d5aefd477dc86a35cc3eb137a8a45534db285/lucene/core/src/test/org/apache/lucene/search/TestConstantScoreScorer.java#L94
... which is fine if you force merge to 1 segment.

Reply:
Oh yes, I understand, thank you :+1: